src/lib.rs

```diff
@@ -129,13 +129,53 @@ impl<T, N> DerefMut for GenericArray<T, N> where N: ArrayLength<T> {
     }
 }
 
+impl<T, N> GenericArray<T, N> where N: ArrayLength<T> {
+    /// map a function over a  slice to a `GenericArray`.
+    /// The length of the slice *must* be equal to the length of the array
+    pub fn map_slice<S, F: Fn(&S) -> T>(s: &[S], f: F) -> GenericArray<T, N> {
+        assert_eq!(s.len(), N::to_usize());
+        map_inner(s, f)
+    }
+    
+    /// map a function over a `GenericArray`.
+    pub fn map<U, F>(self, f: F) -> GenericArray<U, N>
+    where F: Fn(&T) -> U, N: ArrayLength<U> {
+        map_inner(&self, f)
+    }
+}
+
+#[inline]
+fn map_inner<S, F, T, N>(list: &[S], f: F) -> GenericArray<T, N>
+where F: Fn(&S) -> T, N: ArrayLength<T> {
+     unsafe {
+        let mut res : GenericArray<T, N> = std::mem::uninitialized();
+        if let Err(e) = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
```

Comment:
Should panic catching be used in normal cases? I was under the impression it was only supposed to be used at FFI boundaries.

What if you mapped the uninitialized result in another struct with a `Drop` impl that will only run if the array is fully constructed? Something that might look like (warning, bad pseudo code follows):
``` rust
struct ArrayBuilder {
    res: Option<GenericArray<T, N>>,
    len: usize,
}

impl ArrayBuilder {
    fn new() -> Self {
        ArrayBuilder { res: Some(mem::uninitialized()), len: 0 }
    }

    fn append(item: T) {
        ptr::write(&array[len], item);
        self.len += 1;
    }

    fn finish(self) -> GenericArray {
        self.res.take()
    }
}

impl Drop for ArrayBuilder {
    fn drop(&mut self) {
        if self.len != array.len() {
            mem::forget(self.res.take());
        }
    }
}
```

I think an approach like that could givekn us safety without having to catch panics. But I haven't tried to compile or test any of this code :) Curious to hear what you think!